File: 2025_02/leetcode2025-02-27.py
class TextEditor:

    def __init__(self):
        self.string = ""
        self.ptr = 0

    def addText(self, text: str) -> None:
        self.string = self.string[:self.ptr]+text+self.string[self.ptr:]
        self.ptr += len(text)

    def deleteText(self, k: int) -> int:
        l = max(0, self.ptr-k)
        self.string = self.string[:l]+self.string[self.ptr:]
        res = self.ptr-l
        self.ptr = l
        return res

    def cursorLeft(self, k: int) -> str:
        self.ptr = max(0, self.ptr-k)
        return self.string[max(0, self.ptr-10):self.ptr]

    def cursorRight(self, k: int) -> str:
        self.ptr = min(len(self.string), self.ptr+k)
        return self.string[max(0, self.ptr-10):self.ptr]


if __name__ == "__main__":
    te = TextEditor()
    op = ["addText","deleteText","addText","cursorRight","cursorLeft","deleteText","cursorLeft","cursorRight"]
    adds = ["leetcode","practice"]
    nums = [4,3,8,10,2,6]
    a,n = 0,0
    for o in op:
        if o == "addText":
            te.addText(adds[a])
            a += 1
        elif o == "deleteText":
            print(te.deleteText(nums[n]))
            n += 1
        elif o == "cursorRight":
            print(te.cursorRight(nums[n]))
            n += 1
        else:
            print(te.cursorLeft(nums[n]))
            n += 1

File: leetcode2025/2025_02/leetcode2025-02-27.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class TextEditor {
private:
    vector<char> left, right;
public:
    TextEditor() {
    }
    
    void addText(string text) {
        for (char c : text) left.push_back(c);
    }
    
    int deleteText(int k) {
        int res = 0;
        while (k && left.size()>0){
            left.erase(left.end()-1);
            k--;
            res++;
        }
        return res;
    }
    
    string cursorLeft(int k) {
        while (k && left.size()>0){
            char tmp = left[left.size()-1];
            left.erase(left.end()-1);
            right.push_back(tmp);
            k--;
        }
        string res;
        k = 10;
        for (int i=left.size()-1; i>=0&&k>0; i--){
            res = left[i]+res;
            k--;
        }
        return res;
    }
    
    string cursorRight(int k) {
        while (k && right.size()>0){
            char tmp = right[right.size()-1];
            right.erase(right.end()-1);
            left.push_back(tmp);
            k--;
        }
        string res;
        k = 10;
        for (int i=left.size()-1; i>=0&&k>0; i--){
            res = left[i]+res;
            k--;
        }
        return res;
    }
};

int main(){
    TextEditor te;
    vector<string> op = {"addText","deleteText","addText","cursorRight","cursorLeft","deleteText","cursorLeft","cursorRight"};
    vector<string> adds = {"leetcode","practice"};
    vector<int> nums = {4,3,8,10,2,6};
    int a = 0, n = 0;
    for (auto&o : op){
        if (o == "addText"){
            te.addText(adds[a]);
            a++;
        }
        else if (o == "deleteText"){
            cout << te.deleteText(nums[n]) << endl;
            n++;
        }
        else if (o == "cursorRight"){
            cout << te.cursorRight(nums[n]) << endl;
            n++;
        }
        else{
            cout << te.cursorLeft(nums[n]) << endl;
            n++;
        }
    }
}
